### PR TITLE
Bug 1921909: build and use AWS config file for operator and operand

### PR DESCRIPTION
**Pull request by joelddiaz**
In order to support alternative AWS authentication mechanisms beyond just access/secret key authentication:

- [x] Update the operator Deployment to mount in the projected ServiceAccount token (to a default-by-convention location of /var/run/secrets/openshift/serviceaccount).
- [x] Update the AWS S3 driver to respect the new CredentialsRequest Secret field ('credentials'), but keep supporting the old-style format.
- [x] Build and use an AWS config file when creating the AWS S3 client (whether new-style or old-style Secret data).
- [x] Store the AWS credentials config in the image-registry's image-registry-private-configuration Secret.
- [x] Mount the image-registry-private-configuration Secret into the image-registry Deployment PodSpec so that the image-registry can start consuming creds from the Secret instead of only using access/secret key data (requires separate code changes to image-registry).
- [x] Update the image-registry environment variables so that the operator sets the location of the fully-formed credentials file.
- [x] Modify the PodTemplate to always mount in the ServiceAccount projected volume (to a default-by-convention location of /var/run/secrets/openshift/serviceaccount).

With these changes (and the corresponding image-registry support), it should now be possible to use web identity token authentication. This requires:
- setting up an AWS OpenID identity provider that trusts the cluster's OIDC provider
- creating an AWS IAM Role with the permissions requested in this repo's CredentialsRequest
- granting both ServiceAccounts 'registry' and 'cluster-image-registry-operator' to be allowed to assume the AWS IAM Role
- setting up the Secret holding the AWS credentials (installer-cloud-credentials) so that the 'credentials' field contains a working AWS configuration:
```
[default]
role_arn = arn:aws:iam:ACCOUNT_NUM:role/ROLE_NAME
web_identity_token_file = /var/run/secrets/openshift/serviceaccount/token
```

There is still an open question about if/when we move to temporary credentials by default, how the e2e tests should work. They presently re-use the credentials found in the cluster, but when using web identity authentication, it is the Pod that is granted permissions and there are no credentials that can just be grabbed for the various e2e tests.

This is related to https://issues.redhat.com/browse/CO-1255

### CODE REVIEW

**Comment by joelddiaz** [comment]
/hold
this depends on https://github.com/openshift/image-registry/pull/257

File: pkg/storage/s3/s3_test.go
Comment (dmage) [inline pkg/storage/s3/s3_test.go:54]: `GetCloudCredentials` is not set, the test panics because of this

File: pkg/storage/s3/s3_test.go
Comment (dmage) [inline pkg/storage/s3/s3_test.go:170]: This will create a new config file, why do you need this?

File: pkg/storage/s3/s3.go
Comment (dmage) [inline pkg/storage/s3/s3.go:45]: Let's get rid of this type, it confuses people.

File: pkg/storage/s3/s3.go
Comment (dmage) [inline pkg/storage/s3/s3.go:145]: Let's split this function into two:

* updateConfig
* getCredentialsLocation

I think the only reason why we have it as a single function so that people don't forget to call `updateConfig` before using it. Perhaps it should be checked via tests.

That'll help us to get rid of the local type `S3`.

File: pkg/storage/s3/s3.go
Comment (dmage) [inline pkg/storage/s3/s3.go:339]: This is a strange exercise: create a file, read its content, and delete the file.
I don't have suggestions, but there is room for improvement.

File: pkg/storage/s3/s3.go
Comment (dmage) [inline pkg/storage/s3/s3.go:758]: You'll have an empty file in this case (or with partially written data).

File: pkg/storage/s3/s3.go
Comment (dmage) [reply pkg/storage/s3/s3.go:57]: btw, do we need it? We mock listers to provide custom data.

File: pkg/storage/s3/s3.go
Comment (dmage) [inline pkg/storage/s3/s3.go:770]: Can it contain special symbols? Does the value need to be escaped?

**Review by dmage — CHANGES_REQUESTED** [review]
Is there a way to be sure that the operator always deletes temporary files? By reviewing this code I found one place where the operator can create a file and keep it. I'm not sure that I found all such cases. If the operator doesn't delete files, eventually it may run out of inodes and stop working. If we are going to create files, we need somehow verify that the operator doesn't have leaked temporary files after e2e-aws and e2e-aws-operator tests.

**Comment by joelddiaz** [comment]
> Is there a way to be sure that the operator always deletes temporary files? 

One thing we can possibly do is generate the temporary filename once per driver initialization (I think I saw that there would be at most two driver initializations) and keep overwriting the file when creating the AWS session/client.
That does sort of complicate the e2e tests as we will create one temp file for every driver initialization (so perhaps have the e2e tests reach into the driver/struct to clean up the created file...), but it simplifies the actual in-cluster runtime situation.
WDYT?

File: pkg/storage/s3/s3.go
Comment (joelddiaz) [reply pkg/storage/s3/s3.go:57]: :+1:  It was a bit of a mess to use the listers because the in-cluster pieces still need working AWS configs (which look different than the out-of-cluster AWS clients), so I modified it so that the e2e builds it's own AWS client and doesn't rely on running the S3 driver locally to query the state of things.

File: pkg/storage/s3/s3_test.go
Comment (joelddiaz) [reply pkg/storage/s3/s3_test.go:170]: This is some remnants from an earlier approach where the S3 driver would keep only one copy of the AWS config/creds file. Removing it now...

File: pkg/storage/s3/s3.go
Comment (joelddiaz) [reply pkg/storage/s3/s3.go:770]: I can't find a definitive list of allowed characters for the secret key. I do see the occasional '+', '/', '[', ']' and parenthesis, but nothing that would require escaping. I'll update it to use a single Fprintf() with backticks just to be on the safe side.

**Comment by joelddiaz** [comment]
@dmage i've tried to address your PR feedback. let me know (especially the re-working of the S3 driver to get rid of the S3 struct and the way the config state is maintained).

**Comment by joelddiaz** [comment]
fixed merge conflicts from #638 having merged

**Review by ricardomaraschini — CHANGES_REQUESTED** [review]
Looks good, I have left a few comments for clarification. PTAL

File: manifests/01-registry-credentials-request.yaml
Comment (joelddiaz) [reply manifests/01-registry-credentials-request.yaml:34]: I only discovered this while digging into the issues I was having before I realized I needed to add the new s3:ListMultipartUploadParts permission when running with STS credentials ¯\_(ツ)_/¯

File: manifests/07-operator.yaml
Comment (joelddiaz) [reply manifests/07-operator.yaml:94]: This is what makes the entire web of trust work between AWS and the Pod running inside the cluster. You configure an AWS IAM Identity Provider that trusts identities that the OpenShift cluster can validate/sign. So now the Pod can make an iam:STS API call and say "here is my token signed by an OpenID provider that you trust". Then AWS validates that the token is indeed signed by something that it trusts, and (assuming other things check out properly) AWS will hand back temporary credentials.

**Comment by joelddiaz** [comment]
@ricardomaraschini thanks for the review. FYI, I've put in two new commits in the PR addressing your feedback

File: manifests/01-registry-credentials-request.yaml
Comment (ricardomaraschini) [reply manifests/01-registry-credentials-request.yaml:34]: Could you confirm that this permission removal won't affect our calls to `HeadBucketWithContext` even when using access key and secret? Just trying to be on the safe side here :-)

File: manifests/07-operator.yaml
Comment (ricardomaraschini) [reply manifests/07-operator.yaml:94]: Ah, I understand now. Thanks!

File: pkg/storage/s3/s3_test.go
Comment (ricardomaraschini) [inline pkg/storage/s3/s3_test.go:120]: Was this test wrong before? I mean, why do we expect `VirtualHostedStyle` to be true now?

File: manifests/01-registry-credentials-request.yaml
Comment (joelddiaz) [reply manifests/01-registry-credentials-request.yaml:34]: s3:HeadBucket is not a real permission. Manually adding "s3:HeadBucket" to an IAM policy (because you can't select it through the GUI) and viewing the policy in the GUI visual editor shows the following message:
```
 Unrecognized actions
IAM does not recognize one or more actions. The action name might include a typo or might be part of a previewed or custom service.
s3:HeadBucket
```

File: pkg/storage/s3/s3_test.go
Comment (joelddiaz) [reply pkg/storage/s3/s3_test.go:120]: I hope I can explain this :)

The current test in the master branch calls and tests s3.GetConfig(), but the actual s3 package really uses s3.updateConfigAndGetCredentials() (which in turn does call s3.GetConfig()). And in s3.updateConfigAndGetCredentials() there is an extra step where the VirtualHostedStyle is set on the config depending on various values. https://github.com/openshift/cluster-image-registry-operator/blob/master/pkg/storage/s3/s3.go#L124-L129

The restructured code now has the single s3.UpdateEffectiveConfig() call to do the same work that s3.updateConfigAndGetCredentials() used to do. And the new s3.UpdateEffectiveConfig() now does the setting of the VirtualHostedStyle that only s3.updateConfigAndGetCredentials() used to do.

So now the test case in this PR is updated to call/test s3.UpdateEffectiveConfig() and the VirtualHostedStyle value is being set. And that is why I had to update the test case.

I honestly don't know what this value of VirtualHostedStyle accomplishes, and I just tried to preserve the existing behavior. I spent a lot of time trying to figure out why it is necessary to set VirtualHostedStyle, but I don't believe I was able to fully understand it.

As I re-read the code just now to write up this reply, I see that the current s3.updateConfigAndGetCredentials() actually tests the condition for both Config.Region and Config.RegionEndpoint being unset before it goes down the code path to potentially set Config.VirtualHostedStyle, but my restructuring only checks against Region being unset. I *think* it doesn't affect the way things work as this code really only ever runs the first time we are populating the config, but I could change it to more faithfully reproduce what the current code does.

WDYT?

Comment (ricardomaraschini) [reply pkg/storage/s3/s3_test.go:120]: To be honest I also don't fully understand the impacts of VirtualHostedStyle and when we should leverage it and when we don't (it seems like the user can activate it manually as well). I know it has something to do with the way we reach the bucket: with VirtualHostedStyle the bucket is accessible through `<bucketname>.s3.amazon.com` instead of `s3.amazonaws.com/<bucketname>`. That being said I think we can agree that the least we change the current behavior the better, specially because @dmage, that would be able to explain us exactly what that means and in what circumstances we need to use it, is off. BTW, I think that the access without VirtualHostedStyle (for new buckets) has been deprecated: https://aws.amazon.com/blogs/aws/amazon-s3-path-deprecation-plan-the-rest-of-the-story/

Comment (joelddiaz) [reply pkg/storage/s3/s3_test.go:120]: :+1: i put back the condition to match the previous behavior so only if Region and RegionEndpoint were zero-length would it go down that codepath.

**Comment by joelddiaz** [comment]
@dmage @ricardomaraschini can we get a re-review based on the last changes addressing PR feedback?

**Comment by sdodson** [comment]
/retitle Bug 1921909: build and use AWS config file for operator and operand

File: manifests/07-operator.yaml
Comment (ricardomaraschini) [reply manifests/07-operator.yaml:94]: @joelddiaz Isn't this the token for the `cluster-image-registry-operator` service account? If this is the case do we still need to explicitly mount it? As we are not explicitly opting out of `automountServiceAccountToken` the token is already mounted under `/var/run/secrets/kubernetes.io/serviceaccount`. Please clarify.

File: pkg/resource/deployment_test.go
Comment (ricardomaraschini) [inline pkg/resource/deployment_test.go:108]: Can we user a context with a timeout here? If these caches never get ready we won't know exactly what happen other the whole test timing out. Something like below would make it more clear (using a context with Timeout):

```
if !cache.WaitForCacheSync(
		ctx.Done(),
		kubeInformer.Core().V1().ConfigMaps().Informer().HasSynced,
		kubeInformer.Core().V1().Secrets().Informer().HasSynced,
	) {
		t.Fatal("caches not syncing")
	}
```

File: pkg/resource/deployment_test.go
Comment (ricardomaraschini) [reply pkg/resource/deployment_test.go:129]: Yeah, I think this line can be removed.

File: pkg/resource/deployment_test.go
Comment (ricardomaraschini) [inline pkg/resource/deployment_test.go:131]: nit: this should not be necessary, does the test fail if you remove this wait?

File: pkg/resource/podtemplatespec.go
Comment (ricardomaraschini) [inline pkg/resource/podtemplatespec.go:347]: I guess that the answer at https://github.com/openshift/cluster-image-registry-operator/pull/640/files#r566756238 will also solve my question regarding this other service account token :-)

File: pkg/resource/deployment_test.go
Comment (joelddiaz) [reply pkg/resource/deployment_test.go:131]: It's been a while since I touched this, but I believe I added it because I was running into issue otherwise.

File: manifests/07-operator.yaml
Comment (joelddiaz) [reply manifests/07-operator.yaml:94]: The contents of the two ServiceAccounts are slightly different.

Possibly most importantly the keyId field in the kubernetes.io/serviceaccount location doesn't match what is uploaded to the OIDC S3 bucket:

Standard ServiceAccount token:
`sh-4.4$ cat /var/run/secrets/kubernetes.io/serviceaccount/token | awk -F. '{ print $1 }' | base64 -d
{"alg":"RS256","kid":"RjUCNiAxtfGZcTjdVWLvNyksvfhTuQ7dmqAxBrbZIYA"}base64: invalid input
`
Projected ServiceAccount token:
`sh-4.4$ cat /var/run/secrets/openshift/serviceaccount/token | awk -F. '{ print $1 }' | base64 -d
{"alg":"RS256","kid":"lQZjBZuL5Xzq0B6Ih3VQVWFi6zVPE8u7fq_yubth2GQ"}base64: invalid input
`

And looking at the OIDC keys.json, you see the missmatch:

```
[jdiaz@minigoomba ~]$ curl https://jdiaz-stsfull-installer.s3.amazonaws.com/keys.json | jq -r '.keys[].kid'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   917  100   917    0     0   2345      0 --:--:-- --:--:-- --:--:--  2351
lQZjBZuL5Xzq0B6Ih3VQVWFi6zVPE8u7fq_yubth2GQ
```

Looking at the second field of the token there are more differences:

Standard ServiceAccount token:
`sh-4.4$ cat /var/run/secrets/kubernetes.io/serviceaccount/token | awk -F. '{ print $2 }' | base64 -d
{"iss":"kubernetes/serviceaccount","kubernetes.io/serviceaccount/namespace":"openshift-image-registry","kubernetes.io/serviceaccount/secret.name":"cluster-image-registry-operator-token-jbtdj","kubernetes.io/serviceaccount/service-account.name":"cluster-image-registry-operator","kubernetes.io/serviceaccount/service-account.uid":"30278254-d1bc-4631-ad41-777cbef7bb39","sub":"system:serviceaccount:openshift-image-registry:cluster-image-registry-operator"}base64: invalid input`

Projected ServiceAccount:
`sh-4.4$ cat /var/run/secrets/openshift/serviceaccount/token | awk -F. '{ print $2 }' | base64 -d
{"aud":["openshift"],"exp":1611938123,"iat":1611934523,"iss":"https://s3.us-east-1.amazonaws.com/jdiaz-stsfull-installer","kubernetes.io":{"namespace":"openshift-image-registry","pod":{"name":"cluster-image-registry-operator-8cb64687b-bpdp9","uid":"5050f308-9c1d-4c3c-89a4-8c454d0e08f6"},"serviceaccount":{"name":"cluster-image-registry-operator","uid":"30278254-d1bc-4631-ad41-777cbef7bb39"}},"nbf":1611934523,"sub":"system:serviceaccount:openshift-image-registry:cluster-image-registry-operator"}`

Here the issuer field (iss) differs, there is a missing audience (aud), no expiration (exp) or issued-at (iat).

But I do agree that they are fundamentally the same as you can see the serviceaccount.uid match.

Comment (ricardomaraschini) [reply manifests/07-operator.yaml:94]: Thanks for clarifying this aspect. Two distinct tokens indeed :-)

File: pkg/resource/deployment_test.go
Comment (ricardomaraschini) [reply pkg/resource/deployment_test.go:136]: Well, this is not a no go condition, let's just move this forward.

**Comment by ricardomaraschini** [comment]
/unhold

https://github.com/openshift/image-registry/pull/257 is already merged.

**Comment by dmage** [comment]
This is a new feature and there is a strong desire to have it in 4.7.0. As it's a feature, it should be approved (i.e. it should have xxx-approved labels), so please take a look.

/assign @wzheng1
It should be possible to pull/push images into the registry when OpenShift is installed in a regular way. The registry pods should have AWS secret and access key in their `/var/run/secrets/cloud/credentials` file (this file didn't exist before this PR).
It should also be possible to pull/push images when OpenShift installed [with with an STS credential](https://github.com/openshift/enhancements/blob/master/enhancements/cloud-integration/aws/aws-sts-support.md). In this case the registry pods should mention `/var/run/secrets/openshift/serviceaccount/token` in their `/var/run/secrets/cloud/credentials` file.

/assign @bmcelvee
As I can see, we don't need to change anything regarding the registry, all changes are about the installer, cloud-credentials-operator, and CVO. But maybe STS credentials needs somehow be mentioned in the registry docs.

/assign @sferich888
Please take a look.

See also:

* https://issues.redhat.com/browse/CCO-52
* https://github.com/openshift/enhancements/blob/master/enhancements/cloud-integration/aws/aws-sts-support.md
* https://github.com/joelddiaz/enhancements/blob/sts-by-default/enhancements/cloud-integration/aws/aws-sts-by-default.md#installer-manifests

**Comment by bparees** [comment]
@wzheng1 @bmcelvee @sferich888 as we need this prior to codefreeze for 4.7 (feb 5th), please prioritize this as much as possible.

**Comment by joelddiaz** [comment]
> See also:
> 
>     * https://issues.redhat.com/browse/CCO-52
> 
>     * https://github.com/openshift/enhancements/blob/master/enhancements/cloud-integration/aws/aws-sts-support.md
> 
>     * https://github.com/joelddiaz/enhancements/blob/sts-by-default/enhancements/cloud-integration/aws/aws-sts-by-default.md#installer-manifests

I would refer to https://github.com/openshift/cloud-credential-operator/pull/290 for the instructions on how to install in STS-mode (not a generally supported feature FWIW, more of where we intend to take AWS credentials in the future).

For this PR, when a cluster is installed in traditional Mint mode (what CI e2e does), the difference is that the Secret containing the AWS credentials will show an AWS configuration that happens to include the aws_access_key_id and aws_secret_access_key fields.

```
$ oc get secrets -n openshift-image-registry installer-cloud-credentials -o json | jq -r .data.credentials | base64 -d
[default]
aws_access_key_id  = AKACCESSKEYID
aws_secret_access_key = SECRET_ACCESS_KEY
```

And in STS mode, it will reference the AWS IAM Role name to assume, and the location of the ServiceAccount token to AssumeRole() with.
```
$ oc get secrets -n openshift-image-registry installer-cloud-credentials -o json | jq -r .data.credentials | base64 -d
[default]
role_arn = arn:aws:iam::123456789012:role/jdiaz-ststest-image-registry
web_identity_token_file = /var/run/secrets/openshift/serviceaccount/token
```

**Comment by wzheng1** [comment]
@dmage @bparees Get it, QE is working on the test now.

**Comment by gregsheremeta** [comment]
ping 

@dmage can you review the TCs?

Anything else outstanding before this can merge?

File: manifests/07-operator.yaml
Comment (bmcelvee) [reply manifests/07-operator.yaml:94]: Is this automatically set, or does the user need to do anything manually?

Comment (joelddiaz) [reply manifests/07-operator.yaml:94]: The Deployment for the image-registry-operator is permanently changed to start projecting in the ServiceAccount to this /var/run/secrets/openshift/serviceaccount/token location. OpenShift handles signing the token and updating/rotating it as needed automatically. Whether or not we use the ServiceAccount token, it will be there going forward (even for non-AWS clusters).

**Comment by dmage** [comment]
@wzheng1 tests look good, should I add `tc-approved` somewhere?

@gregsheremeta we need 3 approvals: qe-approved from @wzheng1, docs-approved from @bmcelvee, px-approved from @sferich888

File: test/e2e/aws_test.go
Comment (dmage) [inline test/e2e/aws_test.go:708]: @joelddiaz this test fails, you are trying to access this Secret after the operator's config is deleted (at this point the operator is finalizing the config resource and deleting dependant resources).

Comment (joelddiaz) [reply test/e2e/aws_test.go:708]: @dmage IIUC, I suppose I can move grabbing the Secret's content above line 693 where the Configs().Delete() is called.

**Comment by gregsheremeta** [comment]
ping @dmage can you re-review?

**Pull request merged**
